### PR TITLE
#34042 - add teaser option for sending emails

### PR DESCRIPTION
**Pull request by Bonch2001**
Added an option for email notifications to show a teaser instead of the full content, requiring users to click in order to view the full message in Redmine.

### CODE REVIEW

File: app/views/settings/_advanced_messenger_settings.html.erb [hunk 1]
```diff
@@ -13,4 +13,9 @@
     <%= hidden_field_tag 'settings[disable_fix_for_scroll_to_anchor]', 0, :id => nil %>
     <%= check_box_tag 'settings[disable_fix_for_scroll_to_anchor]', 1, Setting.plugin_redmine_advanced_messenger[:disable_fix_for_scroll_to_anchor] == '1' %>
     <em class="info"><%=raw l(:disable_fix_for_scroll_to_anchor_note) %></em>
+</p>
+<p>
+    <label for="settings_notification_mail_content"><%= l(:notification_mail_content) %></label>
+    <%= select_tag 'settings[notifications_mail_option]', options_for_select([['Teaser', 'teaser'], ['Whole note', 'whole_note']], Setting.plugin_redmine_advanced_messenger[:notifications_mail_option]), include_blank: false %>
```
Comment (mircea-negreanu) [inline app/views/settings/_advanced_messenger_settings.html.erb:19]: I think `Teaser` and `Whole note` needs internationalization too.

File: lib/patches/mailer_patch.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        redmine_headers 'Project' => issue.project.identifier,
+                        'Issue-Tracker' => issue.tracker.name,
+                        'Issue-Id' => issue.id,
+                        'Issue-Author' => issue.author.login,
+                        'Issue-Assignee' => assignee_for_header(issue)
+        redmine_headers 'Issue-Priority' => issue.priority.name if issue.priority
+        message_id journal
+        references issue
+
+        @author = journal.user
```
Comment (mircea-negreanu) [inline lib/patches/mailer_patch.rb:29]: I see that the function `issue_edit` and `message_posted` are heavily different from original. Any reason why?

File: lib/patches/mailer_patch.rb [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          @message_teaser = ActionController::Base.helpers.truncate(message.content, length: 100, omission: '...')
+          mail_content = @message_teaser
+        else
+          mail_content = message.content
+        end
+
+        subject = "[#{message.board.project.name} - #{message.board.name} - msg#{message.root.id}] #{message.subject}"
+
+        mail to: user.mail, subject: subject do |format|
+          format.text { render plain: mail_content }
```
Comment (mircea-negreanu) [inline lib/patches/mailer_patch.rb:73]: Why did you need to add those two formats?

File: lib/patches/mailer_patch.rb [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        alias_method :super_issue_edit, :issue_edit
+        alias_method :issue_edit, :patched_issue_edit
+
+        alias_method :super_message_posted, :message_posted
+        alias_method :message_posted, :patched_message_posted
+      end
+    end
+
+    module InstanceMethods
+      def patched_issue_edit(user, journal)
```
Comment (mircea-negreanu) [inline lib/patches/mailer_patch.rb:17]: I'm thinking that a better solution than rewrite the entire function, is to just rewrite the view, and do you modifications there.
If needed, you can also overwrite the function, but just set some instance variables (`@..`) to be accessed in view (if needed). Since the view is and `erb` you can also write ruby there.

Unfortunatally I don't have an example for how to do it. Is possible that when writing a view with the same name in your plugin it just overwrites the default one (from redmine)? You will have to try and see.

File: lib/patches/mailer_patch.rb (hunk 3 above)
Comment (Bonch2001) [reply lib/patches/mailer_patch.rb:73]: The two formats, `format.text` and `format.html`, were added to ensure compatibility with a variety of email clients. Not all clients can render HTML emails, and some users or systems prefer plain text for accessibility or security reasons. By providing both formats, we make sure the email is readable regardless of the client being used, and it also helps reduce the likelihood of the email being flagged as spam. This approach is considered best practice for sending emails in production environments.

File: lib/patches/mailer_patch.rb (hunk 2 above)
Comment (Bonch2001) [reply lib/patches/mailer_patch.rb:29]: I made the changes to `issue_edit` and `message_posted` because I needed to introduce the functionality that controls whether the email sends the teaser or the whole note. My goal was to keep the changes as minimal as possible while ensuring the email content behaves as expected. In my last commit I also refactored the logic for checking the value of the dropdown and computing the email content by moving it into a separate helper function (`mail_content_with_teaser`) to make the code cleaner and easier to maintain. That said, if you think there's a better way to handle this or if the changes feel too extensive, I'm open to revisiting the code further. Let me know if you have any specific ideas or suggestions on how you'd like it adjusted.

File: app/views/settings/_advanced_messenger_settings.html.erb (hunk 1 above)
Comment (Bonch2001) [reply app/views/settings/_advanced_messenger_settings.html.erb]: Good catch! I added internationalization for "Teaser" and "Whole note" after I saw your comment. Thanks for pointing that out!

File: lib/patches/mailer_patch.rb (hunk 4 above)
Comment (Bonch2001) [reply lib/patches/mailer_patch.rb:17]: Thanks for your input! I’ve made the other changes, including the refactoring and internationalization, and committed them.
As for your suggestion to rewrite the view instead of modifying the function, I see your point. Following this approach, I’ll likely need to remove my mailer_patch, which is fine. I’m working on trying this out now and will keep you updated on how it goes.
Let me know if you have any additional thoughts as I proceed with this change.

Comment (Bonch2001) [reply lib/patches/mailer_patch.rb]: I've made the changes you suggested. I removed the logic from the mailer patch and moved it directly into the views. The view now handles whether to show the teaser or the full note based on the setting `notifications_mail_option`. Everything seems to be working well without the need for extra instance variables or patches.

File: app/views/mailer/issue_edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+<hr />
+
+<ul class="journal details">
+<% details_to_strings(@journal_details, false, :only_path => false).each do |string| %>
+  <li><%= string %></li>
+<% end %>
+</ul>
+
+<% if Setting.plugin_redmine_advanced_messenger[:notifications_mail_option] == 'teaser' %>
+  <%= truncate(@journal.notes, length: 100, omission: '...') %>
```
Comment (mircea-negreanu) [inline app/views/mailer/issue_edit.html.erb:14]: Here, I think we should also run `escape_javascript`. See comment in task.

File: app/views/mailer/message_posted.html.erb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+<h1><%= @message.board.project.name %> - <%= @message.board.name %>: <%= link_to(@message.subject, @message_url) %></h1>
+<em><%= @message.author %></em>
+
+<% if Setting.plugin_redmine_advanced_messenger[:notifications_mail_option] == 'teaser' %>
+  <%= truncate(@message.content, length: 100, omission: '...') %>
```
Comment (mircea-negreanu) [inline app/views/mailer/message_posted.html.erb:5]: Here as well.

Comment (Bonch2001) [reply app/views/mailer/message_posted.html.erb]: Done

File: app/views/mailer/issue_edit.html.erb (hunk 5 above)
Comment (Bonch2001) [reply app/views/mailer/issue_edit.html.erb]: Done
